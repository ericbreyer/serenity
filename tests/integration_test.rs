#![cfg(test)]

use anyhow::{bail, Result};
use insta::assert_snapshot;
use test_case::{test_case, test_matrix};

#[test_case("trivial.ser" => matches Ok(1))]
#[test_case("prime_sieve.ser" => matches Ok(8181807856294299570))]
#[test_case("interfaces.ser" => matches Ok(12))]
#[test_case("babbage.ser" => matches Ok(25264))]
#[test_case("linkedlist.ser" => matches Ok(55))]
#[test_case("prime_conspiricy.ser" => matches Ok(7942686168))]
#[test_case("highly_composites.ser" => matches Ok(355168))]
#[test_case("generic.ser" => matches Ok(6))]
#[test_case("generic_func.ser" => matches Ok(540))]
#[test_case("generic_method.ser" => matches Ok(12))]
#[test_case("fib_stream.ser" => matches Ok(1346268))]
#[test_case("mat_fib.ser" => matches Ok(1346268))]
pub fn test_file_run(file: &str) -> Result<usize> {
    let file = format!("tests/test_files/{}", file);

    let mut out = Vec::new();
    let code = serenity::run_file(&file, &mut out)?;

    Ok(code as usize)
}

#[test_matrix(
    ["trivial.ser", "prime_sieve.ser", "interfaces.ser", "babbage.ser", "linkedlist.ser", "prime_conspiricy.ser", "highly_composites.ser", "generic.ser", "generic_func.ser", "generic_method.ser", "fib_stream.ser", "mat_fib.ser"],
    ["scan", "parse", "compile"]
)]
fn test_file_artifacts(file: &str, mode: &str) -> Result<()> {
    let file = format!("tests/test_files/{}", file);
    let result = match mode {
        "scan" => serenity::scan(&file)?
            .iter()
            .fold(String::new(), |acc, x| acc + &format!("{}\n", x)),
        "parse" => serenity::parse(&file)?,
        "compile" => serenity::compile(&file)?,
        _ => bail!("Invalid mode"),
    };

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(format!("{}_{}", file, mode));
    let _g = settings.bind_to_scope();

    assert_snapshot!(result);

    Ok(())
}
