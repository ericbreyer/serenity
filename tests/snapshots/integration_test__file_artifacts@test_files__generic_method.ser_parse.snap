---
source: tests/integration_test.rs
expression: result
---
[line    6] Function Declaration: test_add<T>
[line    6] ╰─── Function Expression:
[line    6]      ├─── Name: test_add
[line    6]      ├─── Return Type: fn[*struct test { a: <T>, b: <T>, }](<T>) -> <T>
[line    6]      ├─── Params:
[line    6]      │    ╰─── t: *struct test { a: <T>, b: <T>, }
[line    6]      ╰─── Body: 
[line    6]           ╰─── Return:
[line    0]                ╰─── Function: test_add_inner
[line    0]                     ╰─── Function Expression:
[line    0]                          ├─── Name: test_add_inner
[line    0]                          ├─── Return Type: <T>
[line    0]                          ├─── Captures:
[line    0]                          │    ╰─── t
[line    0]                          ├─── Params:
[line    0]                          │    ╰─── c: <T>
[line    0]                          ╰─── Body: 
[line    6]                               ╰─── Return:
[line    7]                                    ╰─── Binary: Plus
[line    7]                                         ├─── Binary: Plus
[line    7]                                         │    ├─── Get: "a"
[line    7]                                         │    │    ╰─── Deref: 
[line    7]                                         │    │         ╰─── Variable: t
[line    7]                                         │    ╰─── Get: "b"
[line    7]                                         │         ╰─── Deref: 
[line    7]                                         │              ╰─── Variable: t
[line    7]                                         ╰─── Variable: c
[line   10] Function Declaration: main
[line   10] ╰─── Function Expression:
[line   10]      ├─── Name: main
[line   10]      ├─── Return Type: int
[line   10]      ╰─── Body: 
[line   11]           ├─── Var Declaration: d
[line   12]           ├─── Expression: 
[line   12]           │    ╰─── Assign:
[line   12]           │         ├─── Get: "a"
[line   12]           │         │    ╰─── Variable: d
[line   12]           │         ╰─── Literal: 1
[line   13]           ├─── Expression: 
[line   13]           │    ╰─── Assign:
[line   13]           │         ├─── Get: "b"
[line   13]           │         │    ╰─── Variable: d
[line   13]           │         ╰─── Literal: 2
[line   15]           ├─── Var Declaration: f
[line   16]           ├─── Expression: 
[line   16]           │    ╰─── Assign:
[line   16]           │         ├─── Get: "a"
[line   16]           │         │    ╰─── Variable: f
[line   16]           │         ╰─── Literal: 1
[line   17]           ├─── Expression: 
[line   17]           │    ╰─── Assign:
[line   17]           │         ├─── Get: "b"
[line   17]           │         │    ╰─── Variable: f
[line   17]           │         ╰─── Literal: 2
[line   17]           ╰─── Return:
[line   19]                ╰─── Binary: Plus
[line   19]                     ├─── Call:
[line   19]                     │    ├─── Get: "add"
[line   19]                     │    │    ╰─── Variable: d
[line   19]                     │    ╰─── Literal: 3
[line   19]                     ╰─── Cast: int
[line   19]                          ╰─── Call:
[line   19]                               ├─── Get: "add"
[line   19]                               │    ╰─── Variable: f
[line   19]                               ╰─── Literal: 3
